ext/openssl/openssl.c

```diff
@@ -115,6 +117,16 @@ enum php_openssl_cipher_type {
 	PHP_OPENSSL_CIPHER_DEFAULT = PHP_OPENSSL_CIPHER_RC2_40
 };
 
+/* Add some encoding rules.  This is normally handled through filters
```

Comment:
The file is already a bit monolithic even before your 700+ lines of additions.  How about we put your new stuff in `ext/openssl/openssl-cms.c` to scope it into one place for maintenance purposes?

Reply:
Yeah I was thinking about this already and I think this should be done later and separate the whole file into more parts. For that it will require some internal header as it's using some common functions. Think it's a bit messy to do that only for cms. Also encoding can be used in different parts (it's not CMS specific) so this should be actually common. So I would suggest to leave it for now as it is.

Reply:
I agree that the file is a bit large.  If I look at the functions that are offered they break out into the following categories:
 - X.509 ops, including CSR
 - Public key ops
 - PKCS#12
 - PKCS#7
 - CMS
 - Support macros for PHP
 - Common constants

I like small files myself, but as this is quite a substantial piece of work, I would suggest that it be handled as a separate set of commits.

Reply:
yeah agreed that it should be separate PR.

I would probably go for:
* `openssl.c` - macros, common, init, error...
* `openssl_x509` - x509, csr, spki
* `openssl_pkcs12`
* `openssl_pkcs7`
* `openssl_cms`
* `openssl_cipher` - symmetric encryption and decryption
* `openssl_pkey` - public key ops
* `openssl_rand` - openssl_random_pseudo_bytes

The disadvantage is of course that bugfixes and merging up will result in conflicts but it's probably worth it as there are not too many bugfixes anyway...

Reply:
So long as all of that happens *after* this PR I'm fine with that ;-)